### PR TITLE
optimize File_system.copy_file to avoid 2 userspace copies

### PR DESCRIPTION
**Pull request by domsj**

### CODE REVIEW

File: src/tools/file_system.ml [hunk 1]
```diff
@@ -161,10 +172,27 @@ let copy_file source target ~overwrite ~throttling =
...
-         with_tmp_file tmp_file target
-                       (fun oc -> copy_all ic oc)) >>= fun () ->
+      with_fd
+        source
+        ~flags:[ Unix.O_RDONLY; Unix.O_NONBLOCK; ]
+        ~perm:0
+        (fun fd_in ->
+          with_fd
+            tmp_file
+            ~flags:[ Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC; Unix.O_NONBLOCK; Unix.O_EXCL; ]
```
Comment (toolslive) [inline src/tools/file_system.ml:182]: This is different, compared to the `Lwt_io.with_file` it's replacing (which boils down to:
`[Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC; Unix.O_NONBLOCK]`. The `O_EXCL` makes it refuse copying
over an existing files and symbolic links

File: src/tools/file_system.ml [hunk 2]
```diff
@@ -161,10 +172,27 @@ let copy_file source target ~overwrite ~throttling =
...
+            ~perm:0o644
```
Comment (toolslive) [inline src/tools/file_system.ml:183]: changed compared to `Lwt_io.with_file` but I think this is a better setting than `0o666`

File: src/tools/file_system.ml (hunk 1 above)
Comment (domsj) [reply src/tools/file_system.ml:182]: fyi with_tmp_file above also has the O_EXCL flag

File: src/tools/file_system.ml (hunk 2 above)
Comment (domsj) [reply src/tools/file_system.ml:183]: also stole that part from with_tmp_file :)

**Review by toolslive — APPROVED** [review]

**Pull request merged**